### ISSUE TITLE
D8 dexmerger crash with multidex on bazel 5 and latest version of android build-tools

### ISSUE THREAD

**Issue by mauriciogg** [bazelbuild/bazel#15432]
### Description of the bug:

Getting the following exception when trying to compile an app with multidex enabled:

Android build-tools version 32.0.0

Using the following config:

```
build --define=android_dexmerger_tool=d8_dexmerger
build --define=android_incremental_dexing_tool=d8_dexbuilder
build --incremental_dexing # to overwrite our default

```

`ERROR: /Users/mgalindo/Snapchat/Dev/android/snapchat/apps/mushroom/apk/BUILD.bazel:341:15: Assembling dex files into apps/mushroom/apk/_dx/mushroom-gms/shard1.dex.zip failed: (Exit 1): d8_dexmerger failed: error executing command (from target //apps/mushroom/apk:mushroom-gms)
(cd /Users/mgalindo/Snapchat/Dev/.cache/_bazel_root/1557532aa844cabca730c318accb6923/execroot/__main__ && \
exec env - \
PATH=/bin:/usr/bin:/usr/local/bin \
bazel-out/darwin-opt-exec-2B5CBBC6/bin/external/bazel_tools/tools/android/d8_dexmerger --input bazel-out/darwin-fastbuild/bin/apps/mushroom/apk/_dx/mushroom-gms/shard1.jar.dex.zip --output bazel-out/darwin-fastbuild/bin/apps/mushroom/apk/_dx/mushroom-gms/shard1.dex.zip '--multidex=best_effort')
# Configuration: b7032664c4d11f2857c8e4e63a62a04f5928cd2022f5ca778d14a391d8119217
# Execution platform: @local_config_platform//:host
Error: Cannot fit requested classes in a single dex file (# methods: 146447 > 65536 ; # fields: 90194 > 65536). Try supplying a main-dex list
Merge failed: Compilation failed to complete, position: null
Exception in thread "main" com.android.tools.r8.CompilationFailedException: Merge failed: Compilation failed to complete, position: null
at com.google.devtools.build.android.r8.DexFileMerger.main(DexFileMerger.java:431)`

If instead I build with the following config

```
build --noincremental_dexing
build --define=android_dexmerger_tool=d8_dexmerger
build --define=android_incremental_dexing_tool=d8_dexbuilder
build --define=android_standalone_dexing_tool=d8_compat_dx
```

I get 

```
ERROR: /Users/mgalindo/Snapchat/Dev/android/snapchat/apps/mushroom/apk/BUILD.bazel:341:15: Converting bazel-out/darwin-fastbuild/bin/apps/mushroom/apk/_dx/mushroom-gms/shard7.jar to dex format failed: (Exit 1): d8_compat_dx failed: error executing command (from target //apps/mushroom/apk:mushroom-gms)
(cd /Users/mgalindo/Snapchat/Dev/.cache/_bazel_root/1557532aa844cabca730c318accb6923/execroot/__main__ && \
exec env - \
PATH=/bin:/usr/bin:/usr/local/bin \
bazel-out/darwin-opt-exec-2B5CBBC6/bin/external/androidsdk/d8_compat_dx --dex --multi-dex '--output=bazel-out/darwin-fastbuild/bin/apps/mushroom/apk/_dx/mushroom-gms/shard7.dex.zip' bazel-out/darwin-fastbuild/bin/apps/mushroom/apk/_dx/mushroom-gms/shard7.jar)
# Configuration: 61041b4b718acf898e5adf7a747bac25a6d199ead373309f47e14212161978f5
# Execution platform: @local_config_platform//:host
Error: Could not find or load main class com.android.tools.r8.compatdx.CompatDx
Caused by: java.lang.ClassNotFoundException: com.android.tools.r8.compatdx.CompatDx
```

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

Build an app that will trigger multidex (has more than 64k methods) with bazel 5.0 and above (specifically after this commit https://github.com/bazelbuild/bazel/commit/ceb498fd6b9323a738f0ba628d12a23a161a0671). I don't have an easy example to repro since this triggers on our internal builds and this only reproduce on large apps.

### Which operating system are you running Bazel on?

Mac Os Montery 12.3.1

**Comment by mauriciogg** [comment]
@timpeut fyi

**Comment by nkoroste** [comment]
@ahumesky @ted-xie do you guys see this internally?

**Comment by ted-xie** [comment]
I have personally not seen this. @mauriciogg have you tried a newer Bazel hash? The commit you linked earlier looks like it's from July 2021. We've recently merged a commit that upgrades the D8 version to the latest available in Maven, so I wonder if this is simply a case of "D8 too old".

**Comment by arunkumar9t2** [comment]
@ted-xie 
We also used to see this consistently when using incremental dexing with d8 merger on 5.0+ I think. But as of `6.0.0-pre.20220517.1` we no longer see this and build completes successfully.

**Comment by ted-xie** [comment]
@arunkumar9t2 Glad to hear that the 6.0.0-pre[...] version works for you. I suspect this was a bug in the previous version of D8 dexmerger, and the new version patched a few weeks ago resolves the crash.

**Comment by ted-xie** [comment]
FYI, I have not been able to repro this bug locally yet. Based on my bazel 5.0.0 installation, I also had to add `--nouse_workers_with_dexbuilder` to the build flags, otherwise dexbuilder would complain about unsupported persistent worker mode.

I wrote a code generator to emit an arbitrary number of class methods in a test project, and when I increase the number to well above the 64k limit (66000 in my test), I get a sensical-looking error:

```
java.lang.AssertionError: constant pool has more than 2^16 entries
```

This is with `multidex = "native"` in my android_binary stanza. My bazel version below:

```
$ bazel version
Build label: 5.0.0-pre.20210510.2.2
```
